Lib/cgi.py

```diff
@@ -633,11 +637,23 @@ def read_multi(self, environ, keep_blank_values, strict_parsing):
             if 'content-length' in headers:
                 del headers['content-length']
 
+            # Propagate max_num_fields into the sub class appropriately
+            sub_max_num_fields = self.max_num_fields
+            if sub_max_num_fields is not None:
+                sub_max_num_fields -= len(self.list)
+
             part = klass(self.fp, headers, ib, environ, keep_blank_values,
                          strict_parsing,self.limit-self.bytes_read,
-                         self.encoding, self.errors)
+                         self.encoding, self.errors, sub_max_num_fields)
```

Comment:
@tseaver I'm assuming here that if a user declares a custom `FieldStorageClass` it has to accept the same params as this class. If that isn't a good assumption, aka breaking change, then let me now. So far the other parameters have had the same assumption.